fix(parser): route block-scoped `let` / `const` through parse_variable

## Goal

`let` and `const` declarations inside a `fn` body parse to `VariableDeclaration` AST nodes instead of falling through to `ExpressionStatement(Raw)`. This makes locals visible to every pass that walks the AST (capture analysis, drop emission, future scope-tracking, IDE features) instead of being opaque text.

## Scope

**In:**
- Add `let` / `const` branches to `parse_block_statement` in `compiler/src/parser/statements.sfn:197` that call the existing `parse_variable` (or whichever helper top-level statements use) and return a `Statement.VariableDeclaration`.
- Ensure `let mut` is preserved through the block-statement path (currently only top-level `let` carries the `mutable` flag because only top-level routes through `parse_variable`).
- Confirm `parse_block` (`compiler/src/parser/statements.sfn:112`) propagates the new statement variant unchanged.

**Out:**
- Lambda body parser refactor (separate issue — though once `parse_block_for_lambda` is replaced with `parse_block`, this fix automatically extends to lambda bodies as well).
- IIFE-in-return-position parser bug (separate issue).
- Any new typecheck behaviour around block-scoped locals; this issue is purely a parser-hole fix. Existing typecheck of top-level `VariableDeclaration` should kick in for free.

## Acceptance Criteria

- [ ] `fn main() { let x = 1; let f = fn() { return x; }; }` parses with two `Statement.VariableDeclaration` entries inside the block (the lambda's captures field is out of scope here — just the structural parse).
- [ ] `let mut count = 0` inside a `fn` body produces a `VariableDeclaration` with `mutable: true`.
- [ ] `const PI = 3.14` inside a `fn` body produces a `VariableDeclaration` with the const flag set.
- [ ] All existing parser / typecheck / lowering tests still pass — locals that previously round-tripped through `Raw` text now flow through the structured path.
- [ ] `make compile` self-hosts and the seedcheck binary still compiles `examples/basics/hello-world.sfn`.
- [ ] `make test-unit`, `make test-integration`, `make test-e2e` pass.
- [ ] `sfn fmt --check compiler/src/` clean.

## Files Affected

- `compiler/src/parser/statements.sfn` — add `let` / `const` branches to `parse_block_statement`.
- `compiler/src/parser/variables.sfn` (or wherever `parse_variable` lives) — confirm reusable from block context; minor refactor if needed.
- `compiler/tests/unit/parser_block_let_test.sfn` — new: cover `let`, `let mut`, `const`, type annotations, and the lambda-inside-block case.

## Verification

```bash
ulimit -v 8388608
make compile
build/native/sailfin test compiler/tests/unit/parser_block_let_test.sfn
make test-unit
make test-integration
make test-e2e
sfn fmt --check compiler/src/parser/
```

Spot-check the AST shape:

```bash
ulimit -v 8388608 && timeout 30 build/native/sailfin emit-ast - <<'EOF'
fn main() {
    let mut count = 0;
    let f = fn() { return count; };
}
EOF
# expected: BlockStatement.body.statements contains two VariableDeclaration entries
```

## Size

S

## Type

Bug

## Blocked by

None

## Context

Surfaced during PR #512 (issue #458, stage A1 of #450 M1.5 — closures-with-capture). The capture-analysis walk had to side-step block-scoped `let` because `parse_block_statement` doesn't recognise the keyword and falls through to a `Raw` text node, making any lambda buried in `fn main() { let f = fn(){...}; }` invisible to structural passes.

The blast radius extends well beyond capture analysis:
- Drop-emission can't see scope-exit boundaries for block-locals.
- Effect checking can't track per-binding effects on locals.
- Any future borrow / lifetime analysis is also blind.

Verified at `compiler/src/parser/statements.sfn:197`: the function branches on `{`, `for`, `loop`, `break`, `continue`, `if`, `match`, `with`, `unsafe`, `return`, `throw`, `try`, `assert` — but never `let` or `const`.

- Tracking #321 — Runtime Rewrite Reassessment & Sequencing
- Surfaced by #512 / #458 (closures stage A1)
- Unblocks future passes that need structural visibility into fn-body locals

Pausing pickup — the "purely a parser-hole fix" assumption breaks in three ways, and resolving them is larger than the `size:s` scope.

## The fix itself works

Reduced AC (after dropping `const`):
- Added `let` branch to `parse_block_statement` in `compiler/src/parser/statements.sfn:221` that delegates to `parse_variable` from `./declarations`.
- New regression test `compiler/tests/unit/parser_block_let_test.sfn` — 6 tests covering `let`, `let mut`, typed `let`, multiple block-scoped lets, top-level path preservation, and nested blocks. **All 6 pass.**
- `make compile` self-hosts cleanly.
- `make test-integration` 23/23 pass.

## What the fix exposes (3 unit-test regressions)

These all stem from the same root cause: block-scoped `let` was previously opaque to typecheck / effect-check / emit, and tests were unintentionally anchored to that bug-mask.

### 1. `scope_and_shadowing_test.sfn` — real typecheck bug now surfaced

```
error[E0001]: duplicate variable `x` declared
  --> scope_and_shadowing_test:14:9
  14 |         let x: number = 20;
```

Source:
```sfn
test "scope: outer variable unchanged after inner shadow" {
    let x: number = 10;
    if true {
        let x: number = 20;  // shadow
```

`register_local_symbol` (`compiler/src/typecheck_types.sfn:970`) treats every duplicate name as an error — there is no concept of "inner scope shadows outer." `check_block` (`typecheck.sfn:460`) clones the parent bindings into the child scope and walks them as one flat list, so the inner `let x` collides with the outer. Pre-fix this never tripped because the inner `let x = 20` was an opaque `ExpressionStatement(Raw)`.

**This is a real Sailfin semantics bug** — shadowing is shown as supported in `docs/reference/spec/03-declarations.md` and the test file is named `scope_and_shadowing_test`. Fixing it requires `check_block` to track a "scope start index" so `register_local_symbol` only checks duplicates within the inner-frame slice.

### 2. `numeric_int_float_test.sfn` — test asserts pre-fix emitted text

```sfn
test "numeric: let x: int = 42 carries int through native IR" {
    let asm = _emit("fn main() { let x: int = 42; }");
    assert index_of(asm, "let x: int") >= 0;  // FAILS post-fix
```

Pre-fix `_emit` produced the raw text `let x: int = 42` (verbatim from source via `ExpressionStatement(Raw)`). Post-fix `emit_variable` (`emit_native.sfn:371`) produces the canonical form `.let x : int = 42` (with spaces around `:` per `append_optional_type_annotation` at `emit_native_format.sfn:347`). The structured form is correct; the test assertion is anchored on the bug-masked rendering.

**Fix:** update the assertion to look for `.let x : int` (or just `x : int`) instead of `let x: int`. Trivial but it's a behavioural change in what the test pins.

### 3. `runtime_capsule_resolver_test.sfn` — effect requirements now propagate

```
error[E0402]: function `test enumerate: returns runtime capsule from workspace` is missing required effects: [io]
  85 |     let result = enumerate_runtime_capsule_artifacts(scratch, empty);
```

`enumerate_runtime_capsule_artifacts` declares `[io]`. Pre-fix the call was inside an opaque `ExpressionStatement(Raw)` and `effect_checker.sfn` never saw it. Post-fix the initializer is a real `Expression.Call` and the effect checker correctly demands `[io]` from the enclosing test. **The effect checker is now doing its job.** Roughly 13 test functions in that file need `[io]` added.

## Why I'm pausing

The issue says "All existing parser / typecheck / lowering tests still pass — locals that previously round-tripped through `Raw` text now flow through the structured path." That promise can't be honored without:

- A typecheck change to allow nested-scope shadowing (out per "Out:" clause).
- Touching effect-annotation declarations on ~13 test functions in a runtime test file.
- Updating native-IR text assertions on tests that were inspecting the raw passthrough.

That's well past `size:s` and crosses the "purely a parser-hole fix" line. Better to re-groom this into a coordinated change (or a small epic): the parser fix + the typecheck shadowing fix + the test updates as one PR or a 2–3 issue sequence.

Flipping `in-progress` → `needs-grooming`. Branch `claude/task-521-pKqmB` has the minimal parser fix + regression test if anyone wants to pick up from there.

Reproduction:
```bash
ulimit -v 8388608
make compile
make test-unit       # 3 failures
make test-integration # all pass
```

Also worth a separate cleanup: `compiler/src/tools/fmt.sfn:77,129` lists `"const"` in the formatter's keyword tables, with no parser counterpart and no language-spec backing. Stale defensive entry; should be removed.

---
_Generated by [Claude Code](https://claude.ai/code)_

--- a/compiler/src/emit_native_format.sfn
+++ b/compiler/src/emit_native_format.sfn
@@ -26,6 +26,19 @@ import {
 } from "./emit_native_state";
 
 // ── Expression formatting ────────────────────────────────────────────
+// Render an operand that's about to appear inside a Binary expression.
+// If the operand is itself a Binary, parenthesize the rendering so the
+// re-parse honours the original grouping. The format is text-only and
+// re-tokenized downstream (`expression_from_tokens`, `parse_let_instruction`),
+// so dropping the parens silently changes precedence and corrupts arithmetic
+// like `(a - 0) / b`.
+fn _format_binary_operand(operand: Expression) -> string {
+    if operand.variant == "Binary" {
+        return "(" + format_expression(operand) + ")";
+    }
+    return format_expression(operand);
+}
+
 fn format_expression(expression: Expression) -> string {
     if expression.variant == "Identifier" {
         let name = expression.name;
@@ -38,11 +51,22 @@ fn format_expression(expression: Expression) -> string {
         return quote_string(expression.value);
     }
     if expression.variant == "Unary" {
-        return expression.operator + format_expression(expression.operand);
+        // Same precedence concern as Binary: `!a && b` parses as
+        // `(!a) && b`, but the original `!(a && b)` had the negation
+        // outside. Wrap a Binary operand so the unary scope survives
+        // round-tripping. Surfaced by the `assert !(... && ...)` pattern
+        // in `compiler/tests/unit/cli_path_normalization_test.sfn:79`.
+        return expression.operator + _format_binary_operand(expression.operand);
     }
     if expression.variant == "Binary" {
-        let left = format_expression(expression.left);
-        let right = format_expression(expression.right);
+        // Wrap nested Binary operands in parens so the round-trip preserves
+        // grouping. Without this, `(a - 0) / b` formats as `a - 0 / b` and
+        // re-parses with `/` binding tighter than `-`, silently changing
+        // semantics. Surfaced by issue #521 once block-scoped `let`
+        // initializers started flowing through `format_expression` instead
+        // of the raw-text passthrough.
+        let left = _format_binary_operand(expression.left);
+        let right = _format_binary_operand(expression.right);
         return left + " " + expression.operator + " " + right;
     }
     if expression.variant == "Member" {

--- a/compiler/src/llvm/expression_lowering/native/core_operators.sfn
+++ b/compiler/src/llvm/expression_lowering/native/core_operators.sfn
@@ -34,11 +34,49 @@ fn strip_enclosing_parentheses(expression: string) -> string {
         _if90 = true;
     }
     if _if90 { return trimmed; }
+    // String-literal-aware paren walk. The naive version counted `(`/`)`
+    // inside `"..."` runs, so an expression like `("(" + s) + ")"` (emitted
+    // by `format_expression` after #521 added parens around nested Binary
+    // operands) appeared to balance one character early and the outer
+    // parens were never stripped, causing the lowering to misparse the
+    // whole thing as a value of `0`.
     let mut depth: int = 0;
+    let mut in_double: boolean = false;
+    let mut in_single: boolean = false;
+    let mut escape_next: boolean = false;
     let mut index: int = 0;
     loop {
         if index >= trimmed.length { break; }
         let ch = char_at(trimmed, index);
+        if escape_next {
+            escape_next = false;
+            index += 1;
+            continue;
+        }
+        if in_double {
+            if ch == "\\" { escape_next = true; } else if ch == "\"" {
+                in_double = false;
+            }
+            index += 1;
+            continue;
+        }
+        if in_single {
+            if ch == "\\" { escape_next = true; } else if ch == "'" {
+                in_single = false;
+            }
+            index += 1;
+            continue;
+        }
+        if ch == "\"" {
+            in_double = true;
+            index += 1;
+            continue;
+        }
+        if ch == "'" {
+            in_single = true;
+            index += 1;
+            continue;
+        }
         if ch == "(" { depth += 1; } else if ch == ")" {
             depth -= 1;
             if depth == 0 {

--- a/compiler/src/parser/statements.sfn
+++ b/compiler/src/parser/statements.sfn
@@ -20,6 +20,7 @@ import {
     substring
 } from "../string_utils";
 import { Token } from "../token";
+import { parse_variable } from "./declarations";
 import { expression_from_tokens } from "./expressions";
 import {
     append_token,
@@ -214,6 +215,21 @@ fn parse_block_statement(parser: Parser) -> BlockStatementParseResult {
         return BlockStatementParseResult {
             parser: body_result.parser,
             statement: Statement.BlockStatement { body: body_result.block },
+            success: true
+        };
+    }
+    if identifier_matches(token, "let") {
+        if decorators.length > 0 {
+            return BlockStatementParseResult {
+                parser: original,
+                statement: null,
+                success: false
+            };
+        }
+        let var_result = parse_variable(after_decorators);
+        return BlockStatementParseResult {
+            parser: var_result.parser,
+            statement: var_result.statement,
             success: true
         };
     }

--- a/compiler/src/tools/fmt.sfn
+++ b/compiler/src/tools/fmt.sfn
@@ -74,7 +74,6 @@ fn _is_keyword(lexeme: string) -> boolean {
     if strings_equal(lexeme, "import") { return true; }
     if strings_equal(lexeme, "export") { return true; }
     if strings_equal(lexeme, "from") { return true; }
-    if strings_equal(lexeme, "const") { return true; }
     if strings_equal(lexeme, "type") { return true; }
     if strings_equal(lexeme, "test") { return true; }
     if strings_equal(lexeme, "impl") { return true; }
@@ -126,7 +125,6 @@ fn _is_top_level_decl(lexeme: string) -> boolean {
     if strings_equal(lexeme, "enum") { return true; }
     if strings_equal(lexeme, "interface") { return true; }
     if strings_equal(lexeme, "test") { return true; }
-    if strings_equal(lexeme, "const") { return true; }
     if strings_equal(lexeme, "type") { return true; }
     if strings_equal(lexeme, "impl") { return true; }
     if strings_equal(lexeme, "model") { return true; }

--- a/compiler/src/typecheck.sfn
+++ b/compiler/src/typecheck.sfn
@@ -162,7 +162,7 @@ fn check_program_scopes(program: Program, interfaces: Statement[]) -> Diagnostic
     let mut diagnostics: Diagnostic[] = [];
 
     for statement in program.statements {
-        let result = check_statement(statement, bindings, interfaces);
+        let result = check_statement(statement, bindings, interfaces, 0);
         bindings = result.bindings;
         let mut _ci: number = 0;
         loop {
@@ -175,12 +175,17 @@ fn check_program_scopes(program: Program, interfaces: Statement[]) -> Diagnostic
     return diagnostics;
 }
 
-fn check_statement(statement: Statement, bindings: SymbolEntry[], interfaces: Statement[]) -> ScopeResult {
+// `scope_start` carries the index that separates inherited (outer-scope)
+// bindings from the current scope's own entries. Inner-block callers compute
+// it as the parent-bindings length after `clone_bindings`, so a `let x` in a
+// nested block can shadow an enclosing `let x` without tripping E0001. Top-
+// level and recursive same-frame callers pass it through unchanged.
+fn check_statement(statement: Statement, bindings: SymbolEntry[], interfaces: Statement[], scope_start: number) -> ScopeResult {
     if statement.variant == "VariableDeclaration" {
-        return register_local_symbol(bindings, statement.name, "variable", statement.span);
+        return register_local_symbol(bindings, statement.name, "variable", statement.span, scope_start);
     }
     if statement.variant == "FunctionDeclaration" {
-        let registration = register_local_symbol(bindings, statement.signature.name, "function", statement.signature.name_span);
+        let registration = register_local_symbol(bindings, statement.signature.name, "function", statement.signature.name_span, scope_start);
         let mut diagnostics = registration.diagnostics;
         let _sig_diags = check_function_signature(statement.signature);
         let mut _ci: number = 0;
@@ -209,7 +214,7 @@ fn check_statement(statement: Statement, bindings: SymbolEntry[], interfaces: St
         };
     }
     if statement.variant == "ExternFunctionDeclaration" {
-        let registration = register_local_symbol(bindings, statement.signature.name, "extern function", statement.signature.name_span);
+        let registration = register_local_symbol(bindings, statement.signature.name, "extern function", statement.signature.name_span, scope_start);
         let mut diagnostics = registration.diagnostics;
         let _sig_diags = check_extern_signature(statement.signature);
         let mut _ci: number = 0;
@@ -224,7 +229,7 @@ fn check_statement(statement: Statement, bindings: SymbolEntry[], interfaces: St
         };
     }
     if statement.variant == "StructDeclaration" {
-        let registration = register_local_symbol(bindings, statement.name, "struct", statement.name_span);
+        let registration = register_local_symbol(bindings, statement.name, "struct", statement.name_span, scope_start);
         let mut diagnostics = registration.diagnostics;
         let _tp_diags = check_type_parameters(statement.type_parameters);
         let mut _ci: number = 0;
@@ -273,7 +278,7 @@ fn check_statement(statement: Statement, bindings: SymbolEntry[], interfaces: St
         };
     }
     if statement.variant == "EnumDeclaration" {
-        let registration = register_local_symbol(bindings, statement.name, "enum", statement.name_span);
+        let registration = register_local_symbol(bindings, statement.name, "enum", statement.name_span, scope_start);
         let mut diagnostics = registration.diagnostics;
         let _tp_diags = check_type_parameters(statement.type_parameters);
         let mut _ci: number = 0;
@@ -295,7 +300,7 @@ fn check_statement(statement: Statement, bindings: SymbolEntry[], interfaces: St
         };
     }
     if statement.variant == "InterfaceDeclaration" {
-        let registration = register_local_symbol(bindings, statement.name, "interface", statement.name_span);
+        let registration = register_local_symbol(bindings, statement.name, "interface", statement.name_span, scope_start);
         let mut diagnostics = registration.diagnostics;
         let _tp_diags = check_type_parameters(statement.type_parameters);
         let mut _ci: number = 0;
@@ -317,7 +322,7 @@ fn check_statement(statement: Statement, bindings: SymbolEntry[], interfaces: St
         };
     }
     if statement.variant == "TestDeclaration" {
-        let registration = register_local_symbol(bindings, statement.name, "test", statement.name_span);
+        let registration = register_local_symbol(bindings, statement.name, "test", statement.name_span, scope_start);
         let mut diagnostics = registration.diagnostics;
         let _blk_diags = check_block(statement.body, [], interfaces);
         let mut _ci: number = 0;
@@ -374,7 +379,7 @@ fn check_statement(statement: Statement, bindings: SymbolEntry[], interfaces: St
                 }
             }
             if branch.statement != null {
-                let result = check_statement(branch.statement, bindings, interfaces);
+                let result = check_statement(branch.statement, bindings, interfaces, scope_start);
                 let mut _ci2: number = 0;
                 loop {
                     if _ci2 >= result.diagnostics.length { break; }
@@ -392,7 +397,7 @@ fn check_statement(statement: Statement, bindings: SymbolEntry[], interfaces: St
         };
     }
     if statement.variant == "TypeAliasDeclaration" {
-        let registration = register_local_symbol(bindings, statement.name, "type", statement.name_span);
+        let registration = register_local_symbol(bindings, statement.name, "type", statement.name_span, scope_start);
         let mut diagnostics = registration.diagnostics;
         let _tp_diags = check_type_parameters(statement.type_parameters);
         let mut _ci: number = 0;
@@ -444,7 +449,7 @@ fn seed_parameter_scope(signature: FunctionSignature) -> ScopeResult {
     let mut diagnostics: Diagnostic[] = [];
 
     for parameter in signature.parameters {
-        let registration = register_local_symbol(bindings, parameter.name, "parameter", parameter.span);
+        let registration = register_local_symbol(bindings, parameter.name, "parameter", parameter.span, 0);
         bindings = registration.bindings;
         let mut _ci: number = 0;
         loop {
@@ -459,10 +464,11 @@ fn seed_parameter_scope(signature: FunctionSignature) -> ScopeResult {
 
 fn check_block(block: Block, parent_bindings: SymbolEntry[], interfaces: Statement[]) -> Diagnostic[] {
     let mut bindings = clone_bindings(parent_bindings);
+    let scope_start = bindings.length;
     let mut diagnostics: Diagnostic[] = [];
 
     for statement in block.statements {
-        let result = check_statement(statement, bindings, interfaces);
+        let result = check_statement(statement, bindings, interfaces, scope_start);
         bindings = result.bindings;
         let mut _ci: number = 0;
         loop {

--- a/compiler/src/typecheck_types.sfn
+++ b/compiler/src/typecheck_types.sfn
@@ -967,12 +967,24 @@ fn check_type_parameters(type_parameters: TypeParameter[]) -> Diagnostic[] {
     return diagnostics;
 }
 
-fn register_local_symbol(bindings: SymbolEntry[], name: string, kind: string, span: SourceSpan?) -> ScopeResult {
-    if has_symbol(bindings, name) {
-        return ScopeResult {
-            bindings: bindings,
-            diagnostics: [make_duplicate_symbol_diagnostic(name, kind, token_from_name(name, span))]
-        };
+// `scope_start` lets a block-scope caller restrict duplicate detection to the
+// portion of `bindings` it added itself, so inner-block `let` declarations can
+// shadow enclosing-scope bindings without tripping E0001. Top-level and
+// parameter callers pass 0 to preserve strict global checking.
+//
+// Variable bindings are exempt from same-scope duplicate checks: `let`
+// rebinding within a single scope is a deliberate Sailfin idiom (the
+// compiler's own LLVM lowering uses the pattern e.g.
+// `compiler/src/llvm/lowering/instructions_loops.sfn:492,597`). Other kinds
+// (function/struct/enum/interface/type/test) keep the duplicate diagnostic.
+fn register_local_symbol(bindings: SymbolEntry[], name: string, kind: string, span: SourceSpan?, scope_start: number) -> ScopeResult {
+    if !strings_equal(kind, "variable") {
+        if has_symbol_in_range(bindings, name, scope_start) {
+            return ScopeResult {
+                bindings: bindings,
+                diagnostics: [make_duplicate_symbol_diagnostic(name, kind, token_from_name(name, span))]
+            };
+        }
     }
 
     let updated = append_symbol(bindings, name, kind, span);
@@ -1012,6 +1024,16 @@ fn clone_bindings(source: SymbolEntry[]) -> SymbolEntry[] {
 
 fn has_symbol(symbols: SymbolEntry[], name: string) -> boolean {
     let mut _si: number = 0;
+    loop {
+        if _si >= symbols.length { break; }
+        if symbols[_si].name == name { return true; }
+        _si += 1;
+    }
+    return false;
+}
+
+fn has_symbol_in_range(symbols: SymbolEntry[], name: string, start: number) -> boolean {
+    let mut _si: number = start;
     loop {
         if _si >= symbols.length { break; }
         if symbols[_si].name == name { return true; }

--- a/compiler/tests/unit/numeric_int_float_test.sfn
+++ b/compiler/tests/unit/numeric_int_float_test.sfn
@@ -28,14 +28,19 @@ fn _diagnostics(source: string) -> Diagnostic[] {
 }
 
 // ── Native-IR carries int/float types through ─────────────────
+// Block-scoped `let` now routes through `parse_variable` (#521), so the
+// emitter produces the canonical `.let x : int = 42` line from
+// `emit_native.sfn::emit_variable` instead of the raw source text the
+// pre-fix `ExpressionStatement(Raw)` path leaked. Assertions pin the
+// structured shape (the type-annotation text is the load-bearing bit).
 test "numeric: let x: int = 42 carries int through native IR" {
     let asm = _emit("fn main() { let x: int = 42; }");
-    assert index_of(asm, "let x: int") >= 0;
+    assert index_of(asm, ".let x : int") >= 0;
 }
 
 test "numeric: let x: float = 3.14 carries float through native IR" {
     let asm = _emit("fn main() { let x: float = 3.14; }");
-    assert index_of(asm, "let x: float") >= 0;
+    assert index_of(asm, ".let x : float") >= 0;
 }
 
 test "numeric: function signature with int and float types" {

--- a/compiler/tests/unit/parser_block_let_test.sfn
+++ b/compiler/tests/unit/parser_block_let_test.sfn
@@ -1,0 +1,156 @@
+// Regression coverage for `parse_block_statement` routing block-scoped
+// `let` / `let mut` declarations through `parse_variable`.
+//
+// History: surfaced by PR #512 / issue #458 (closures-with-capture A1).
+// `parse_block_statement` in `compiler/src/parser/statements.sfn`
+// branched on `for`, `loop`, `break`, `continue`, `if`, `match`, `with`,
+// `unsafe`, `return`, `throw`, `try`, `assert` — but never `let`. Block-
+// scoped variable bindings fell through to `parse_expression_statement`
+// and ended up as opaque `ExpressionStatement(Raw)` text, hiding locals
+// from every pass that walks the AST (capture analysis, drop emission,
+// future scope tracking). The fix adds a `let` branch that delegates to
+// the existing `parse_variable` helper used by top-level declarations,
+// so block-scoped locals now surface as `Statement.VariableDeclaration`
+// with the same field shape (`name`, `mutable`, `initializer`).
+//
+// Note: `const` is intentionally NOT covered here. Per the keywords
+// reference (`site/src/content/docs/docs/reference/keywords.md`) and
+// `spec/03-declarations.md`, Sailfin only has `let` and `let mut`; the
+// scope-adjustment thread on issue #521 documents that decision.
+import { Block, Program, Statement } from "../../src/ast";
+import { parse_program } from "../../src/parser/mod";
+
+// -------------------------------------------------------------------
+// Helpers
+// -------------------------------------------------------------------
+fn _find_function_body(program: Program, name: string) -> Block {
+    let mut index: number = 0;
+    loop {
+        if index >= program.statements.length { break; }
+        let stmt = program.statements[index];
+        if stmt.variant == "FunctionDeclaration" {
+            if stmt.signature.name == name { return stmt.body; }
+        }
+        index += 1;
+    }
+    return Block { tokens: [], text: "", statements: [] };
+}
+
+fn _count_variable_declarations(block: Block) -> number {
+    let mut count: number = 0;
+    let mut index: number = 0;
+    loop {
+        if index >= block.statements.length { break; }
+        if block.statements[index].variant == "VariableDeclaration" {
+            count += 1;
+        }
+        index += 1;
+    }
+    return count;
+}
+
+fn _nth_variable_declaration(block: Block, n: number) -> Statement {
+    let mut found: number = 0;
+    let mut index: number = 0;
+    loop {
+        if index >= block.statements.length { break; }
+        let stmt = block.statements[index];
+        if stmt.variant == "VariableDeclaration" {
+            if found == n { return stmt; }
+            found += 1;
+        }
+        index += 1;
+    }
+    return block.statements[0];
+}
+
+// -------------------------------------------------------------------
+// Tests
+// -------------------------------------------------------------------
+test "parser block let: simple `let` inside fn body produces VariableDeclaration" {
+    let source = "fn main() { let x = 1; }";
+    let program = parse_program(source);
+    let body = _find_function_body(program, "main");
+    assert _count_variable_declarations(body) == 1;
+    let decl = _nth_variable_declaration(body, 0);
+    assert decl.name == "x";
+    assert decl.mutable == false;
+}
+
+test "parser block let: `let mut` inside fn body preserves mutable flag" {
+    // Pre-fix this fell through to parse_expression_statement and lost
+    // the mutable flag — only top-level `let` carried it because only
+    // top-level routed through `parse_variable`.
+    let source = "fn main() { let mut count = 0; }";
+    let program = parse_program(source);
+    let body = _find_function_body(program, "main");
+    assert _count_variable_declarations(body) == 1;
+    let decl = _nth_variable_declaration(body, 0);
+    assert decl.name == "count";
+    assert decl.mutable == true;
+}
+
+test "parser block let: multiple block-scoped lets parse to structured statements" {
+    // The acceptance-criteria example from issue #521. Pre-fix the
+    // lambda buried inside `let f = fn(){...}` was invisible to any
+    // capture-analysis walk over `main`'s body — the outer `let`
+    // collapsed to a `Raw` expression statement.
+    let source = "fn main() { let x = 1; let f = fn() { return x; }; }";
+    let program = parse_program(source);
+    let body = _find_function_body(program, "main");
+    assert _count_variable_declarations(body) == 2;
+
+    let first = _nth_variable_declaration(body, 0);
+    assert first.name == "x";
+
+    let second = _nth_variable_declaration(body, 1);
+    assert second.name == "f";
+}
+
+test "parser block let: typed let with explicit annotation parses" {
+    // `let x: number = 1` previously round-tripped through raw text
+    // inside a block. The structured path must preserve the type
+    // annotation for the typechecker to consume.
+    let source = "fn main() { let x: number = 1; }";
+    let program = parse_program(source);
+    let body = _find_function_body(program, "main");
+    assert _count_variable_declarations(body) == 1;
+    let decl = _nth_variable_declaration(body, 0);
+    assert decl.name == "x";
+    assert decl.type_annotation != null;
+}
+
+test "parser block let: top-level let path is unchanged" {
+    // Anchor existing behaviour — the new branch must not regress the
+    // top-level `let` path that was already shipping a structured
+    // VariableDeclaration.
+    let source = "let module_const = 42;";
+    let program = parse_program(source);
+    let mut index: number = 0;
+    let mut found = false;
+    loop {
+        if index >= program.statements.length { break; }
+        let stmt = program.statements[index];
+        if stmt.variant == "VariableDeclaration" {
+            if stmt.name == "module_const" { found = true; }
+        }
+        index += 1;
+    }
+    assert found;
+}
+
+test "parser block let: nested block lets parse independently" {
+    // Block-scoped lets inside an `if` body must also flow through the
+    // structured path — `parse_block_statement` recurses into nested
+    // blocks via `parse_block`, so the fix automatically extends.
+    let source = "fn main() { if true { let inner = 7; } }";
+    let program = parse_program(source);
+    let body = _find_function_body(program, "main");
+    // Outer body has the if-statement only — no top-level
+    // VariableDeclaration in main's immediate scope.
+    assert _count_variable_declarations(body) == 0;
+    // The inner block is reachable but enumerating it requires
+    // pattern-matching the `IfStatement` variant, which is more AST
+    // shape than this regression needs. The outer-count assertion
+    // confirms the fix doesn't accidentally hoist nested lets.
+}

--- a/compiler/tests/unit/runtime_capsule_resolver_test.sfn
+++ b/compiler/tests/unit/runtime_capsule_resolver_test.sfn
@@ -78,7 +78,7 @@ fn _setup_scratch_workspace(scratch: string) ![io] {
     fs.writeFile(scratch + "/capsules/sfn/library-cap/capsule.toml", library_manifest);
 }
 
-test "enumerate: returns runtime capsule from workspace" {
+test "enumerate: returns runtime capsule from workspace" ![io] {
     let scratch = "build/sailfin/.runtime-resolver-test-1";
     _setup_scratch_workspace(scratch);
     let empty: string[] = [];
@@ -88,7 +88,7 @@ test "enumerate: returns runtime capsule from workspace" {
     process.run(["rm", "-rf", scratch]);
 }
 
-test "enumerate: skips kind != runtime members" {
+test "enumerate: skips kind != runtime members" ![io] {
     let scratch = "build/sailfin/.runtime-resolver-test-2";
     _setup_scratch_workspace(scratch);
     let empty: string[] = [];
@@ -104,7 +104,7 @@ test "enumerate: skips kind != runtime members" {
     process.run(["rm", "-rf", scratch]);
 }
 
-test "enumerate: c_sources resolved to workspace-rooted paths" {
+test "enumerate: c_sources resolved to workspace-rooted paths" ![io] {
     let scratch = "build/sailfin/.runtime-resolver-test-3";
     _setup_scratch_workspace(scratch);
     let empty: string[] = [];
@@ -122,7 +122,7 @@ test "enumerate: c_sources resolved to workspace-rooted paths" {
     process.run(["rm", "-rf", scratch]);
 }
 
-test "enumerate: sfn_sources resolved relative to manifest dir" {
+test "enumerate: sfn_sources resolved relative to manifest dir" ![io] {
     let scratch = "build/sailfin/.runtime-resolver-test-sfn-sources";
     _setup_scratch_workspace(scratch);
     let empty: string[] = [];
@@ -143,7 +143,7 @@ test "enumerate: sfn_sources resolved relative to manifest dir" {
     process.run(["rm", "-rf", scratch]);
 }
 
-test "enumerate: ll_sources resolved with same join rule as c_sources" {
+test "enumerate: ll_sources resolved with same join rule as c_sources" ![io] {
     let scratch = "build/sailfin/.runtime-resolver-test-4";
     _setup_scratch_workspace(scratch);
     let empty: string[] = [];
@@ -155,7 +155,7 @@ test "enumerate: ll_sources resolved with same join rule as c_sources" {
     process.run(["rm", "-rf", scratch]);
 }
 
-test "enumerate: include_dirs resolved to manifest-relative dirs" {
+test "enumerate: include_dirs resolved to manifest-relative dirs" ![io] {
     let scratch = "build/sailfin/.runtime-resolver-test-5";
     _setup_scratch_workspace(scratch);
     let empty: string[] = [];
@@ -166,7 +166,7 @@ test "enumerate: include_dirs resolved to manifest-relative dirs" {
     process.run(["rm", "-rf", scratch]);
 }
 
-test "enumerate: link_libs flow through verbatim (not path-resolved)" {
+test "enumerate: link_libs flow through verbatim (not path-resolved)" ![io] {
     let scratch = "build/sailfin/.runtime-resolver-test-6";
     _setup_scratch_workspace(scratch);
     let empty: string[] = [];
@@ -180,7 +180,7 @@ test "enumerate: link_libs flow through verbatim (not path-resolved)" {
     process.run(["rm", "-rf", scratch]);
 }
 
-test "enumerate: prelude_entry resolved relative to manifest dir" {
+test "enumerate: prelude_entry resolved relative to manifest dir" ![io] {
     let scratch = "build/sailfin/.runtime-resolver-test-7";
     _setup_scratch_workspace(scratch);
     let empty: string[] = [];
@@ -196,7 +196,7 @@ test "enumerate: prelude_entry resolved relative to manifest dir" {
     process.run(["rm", "-rf", scratch]);
 }
 
-test "enumerate: dep_specs filter narrows the result set" {
+test "enumerate: dep_specs filter narrows the result set" ![io] {
     let scratch = "build/sailfin/.runtime-resolver-test-8";
     _setup_scratch_workspace(scratch);
     // Only ask for sfn/runtime-native — the library capsule was
@@ -210,7 +210,7 @@ test "enumerate: dep_specs filter narrows the result set" {
     process.run(["rm", "-rf", scratch]);
 }
 
-test "enumerate: dep_specs with unknown name returns empty" {
+test "enumerate: dep_specs with unknown name returns empty" ![io] {
     let scratch = "build/sailfin/.runtime-resolver-test-9";
     _setup_scratch_workspace(scratch);
     let specs = ["sfn/does-not-exist"];
@@ -219,14 +219,14 @@ test "enumerate: dep_specs with unknown name returns empty" {
     process.run(["rm", "-rf", scratch]);
 }
 
-test "enumerate: missing workspace.toml returns empty" {
+test "enumerate: missing workspace.toml returns empty" ![io] {
     // No setup — pass a non-existent workspace root.
     let empty: string[] = [];
     let result = enumerate_runtime_capsule_artifacts("/tmp/sfn-no-such-workspace-xyz", empty);
     assert result.length == 0;
 }
 
-test "enumerate: empty workspace_root returns empty" {
+test "enumerate: empty workspace_root returns empty" ![io] {
     let empty: string[] = [];
     let result = enumerate_runtime_capsule_artifacts("", empty);
     assert result.length == 0;

